### PR TITLE
Sundry mode and kind syntax

### PR DESCRIPTION
**Pull request by dvulakh**
### Summary
This PR adds support for:
- New shapes of jkinds (product kinds and the `_`, `mod`, `with`, and `kind_of` syntaxes; parsed as in flambda's [#3076]( https://github.com/ocaml-flambda/flambda-backend/pull/3076))
- `kind_abbrev_` in structures and signatures
- `include S @@ mode` in signatures
- The `stack_` keyword in expressions

It also fixes existing bugs around:
- arity-aware representation of function syntax
- erasing modes

### Parser
We update the **standard** and **jane** parsers using the import script to match the state of flambda immediately before the 5.2 merge, _with the exception that [#3076]( https://github.com/ocaml-flambda/flambda-backend/pull/3076) is manually patched in_. That PR went in after the 5.2 merge, but updating `ocamlformat` to match 5.2's parsetree is beyond the scope of this PR.

We change the **extended** parser & lexer to support syntax added since it was last updated, adding paths for erasing janestreet-specific syntax where appropriate. We also add previously neglected erasing to the existing mode parsing logic.

### Normalization
Most changes to standard parsetree normalization logic directly correspond to new erasable (or newly erased) syntax. We also fix normalization of arrows to account for the parser invariant that `Pparam_newtype` cannot be the only 'parameter' to a function.

### Contexts
We add several new contexts, which we need in order to correctly format:
- jkinds, which might need to be parenthesized (depending on the surrounding jkind)
- types (which can now appear inside kinds)

### Misc
In signatures, the (compiler's) parser fails to associate doc comments that follow `include` items. This hasn't caused problems for `ocamlformat` in the past, because the parser instead associated the comment in:
```ocaml
include S (** doc *)
```
with `S` (and `ocamlformat` handled this behavior fine). Modes don't have slots for attributes, however, so now the doc comment in:
```ocaml
inclue S @@ mode (** doc *)
```
disappears from the AST.

We should fix this is the parser upstream, but for now we just extend an existing hack to ensure `ocamlformat` never moves doc comments attached to something in an `include` to after an `include` with modalities.

Jkinds are formatted with more parantheses than strictly necessary for readability. For example,
```ocaml
type t : j & k mod m
```
is rewritten to
```ocaml
type t : (j & k) mod m
```

### Testing
For the most part, each new syntactic construct receives a new test file (together with a corresponding `-erased` test file). Notably, `test/passing/tests/layout_annotation_composed.ml.js-ref` contains generated tests of kind formatting.

We've also tested on the codebase internally (though this is mostly to catch regressions in formatting of existing syntax).

### CODE REVIEW

File: test/passing/tests/layout_annotation_mod.ml.js-ref [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+  | T :
+      ((* 1 *) 'a : value mod mode) 'b (* 2 *) ('c : (* 3 *) float64 mod mode) 'd.
+      (* 4 *) 'a * 'b * 'c * 'd
+      -> t
+
+type t =
+  | T :
+      ((* 1 *) 'a : value mod mode) 'b (* 2 *) ('c : (* 3 *) float64 (* 4 *) mod mode) 'd.
+      (* 5 *)
+
```
Comment (ccasin) [inline test/passing/tests/layout_annotation_mod.ml.js-ref]: This empty line seems weird and would be nice to eliminate if easy.

File: test/passing/tests/layout_annotation_with.ml.js-ref [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+type t =
+  | T :
+      ((* 1 *) 'a : value with type_)
+      'b
+      (* 2 *)
+      ('c : (* 3 *) float64 (* 4 *) with type_)
+      'd.
+      (* 5 *)
+
```
Comment (ccasin) [inline test/passing/tests/layout_annotation_with.ml.js-ref]: Here again a slightly surprising empty line.

File: vendor/parser-jane/repatch.sh [hunk 3]
```diff
@@ -12,24 +12,31 @@ cd $(dirname $0)
...
+        diff3 -m -L old-tip -L old-compiler -L new-compiler \
+              $dir/$file \
+              parser-jane-old/for-$dir/$file \
+              parser-jane/for-$dir/$file \
+              | sed '/^<<<<<<< old-compiler/,/^>>>>>>> new-compiler/{/^<<<<<<< old-compiler/,/^=======/d;/^>>>>>>> new-compiler/d}' \
+              > $dir/$file.tmp \
+              || true
+
+        mv $dir/$file.tmp $dir/$file
+    done
```
Comment (ccasin) [inline vendor/parser-jane/repatch.sh]: This seems like an improvement, but should we update any documentation correspondingly (e.g., `vendor/parser-jane/README.md`).  (Or maybe we should just pull this out into a separate PR - based on the below I don't think you intended to commit it)

File: vendor/parser-jane/repatch.sh [hunk 4]
```diff
@@ -12,24 +12,31 @@ cd $(dirname $0)
...
+
+    hg revert $dir/{jbuild,.hgignore.in,LICENSE}
```
Comment (ccasin) [inline vendor/parser-jane/repatch.sh]: This looks wrong.

File: vendor/parser-jane/repatch.sh [hunk 5]
```diff
@@ -12,24 +12,31 @@ cd $(dirname $0)
...
 done
+
+cp parser-standard/asttypes.ml{i,}
+cp parser-standard/parsetree.ml{i,}
```
Comment (ccasin) [inline vendor/parser-jane/repatch.sh]: I'm confused about this - the old version didn't do this, and I don't see these files in the repo.

File: lib/Ast.ml [hunk 6]
```diff
@@ -1802,14 +1836,16 @@ end = struct
...
-          | Sig _ | Str _ | Clf _ | Ctf _ | Rep | Mb _ | Md _ )
+          ( Pld _ | Top | Tli _ | Typ _ | Tyv _ | Kab _ | Cty _ | Pat _
+          | Mty _ | Mod _ | Sig _ | Str _ | Clf _ | Ctf _ | Rep | Mb _ | Md _
+          | Jkd _ )
       ; ast=
-          ( Pld _ | Top | Tli _ | Pat _ | Exp _ | Cl _ | Mty _ | Mod _
-          | Sig _ | Str _ | Clf _ | Ctf _ | Rep | Mb _ | Md _ ) } ->
+          ( Pld _ | Top | Tli _ | Tyv _ | Kab _ | Pat _ | Exp _ | Cl _
+          | Mty _ | Mod _ | Sig _ | Str _ | Clf _ | Ctf _ | Rep | Mb _ | Md _
+          | Jkd _ ) } ->
```
Comment (ccasin) [inline lib/Ast.ml:1848]: I'm a little surprised we got away here without needing to say anything about the relative precedence of `&` and the other kind operators. The tests give me confidence this works, but do you have any intuition for why we didn't need to do that?

File: lib/Exposed.ml [hunk 7]
```diff
@@ -117,8 +117,9 @@ module Right = struct
     | Pstr_exception te -> type_exception te
     | Pstr_primitive vd -> value_description vd
     | Pstr_module _ | Pstr_recmodule _ | Pstr_modtype _ | Pstr_open _
-     |Pstr_class _ | Pstr_class_type _ | Pstr_include _ | Pstr_attribute _
-     |Pstr_extension _ | Pstr_value _ | Pstr_eval _ ->
+     |Pstr_kind_abbrev _ | Pstr_class _ | Pstr_class_type _
```
Comment (ccasin) [inline lib/Exposed.ml]: I think this (and the corresponding signature_item case below) needs to consider `kind_of`.  A kind abbrev can end in a type, and that needs similar handling to the `Pstr_type` case above, when the decl has a manifest.

File: lib/Fmt_ast.ml [hunk 8]
```diff
@@ -811,21 +775,97 @@ and type_constr_and_body c xbody =
...
+    | Kind_of type_ ->
+        let type_fmt =
+          fmt_core_type c ~box:true (sub_typ ~ctx:inner_ctx type_)
+        in
+        (false, fmt "kind_of_@ " $ Cmts.fmt_within c loc $ type_fmt)
+    | Product kinds ->
+        let parens =
+          match ctx with
+          | Jkd (Product _ | Mod _ | With _) -> true
+          | _ -> false
```
Comment (ccasin) [inline lib/Fmt_ast.ml:875]: Let's get rid of this catch-all case so we are forced to consider any new kinds.

File: lib/Fmt_ast.ml [hunk 9]
```diff
@@ -4182,30 +4241,44 @@ and fmt_signature_item c ?ext {ast= si; _} =
...
+  | Psig_include
+      ({pincl_mod; pincl_attributes; pincl_loc; pincl_kind}, modalities) ->
       update_config_maybe_disabled c pincl_loc pincl_attributes
       @@ fun c ->
       let doc_before, doc_after, atrs =
-        let force_before = not (Mty.is_simple pincl_mod) in
+        let force_before =
+          (* CR modes: As a temporary hack, we force doc comments before
+             [include S @@ mode], because the parsetree doesn't currently
+             have attribute slots that support placing them after. *)
```
Comment (ccasin) [inline lib/Fmt_ast.ml]: I think we determined the problem is not really the lack of attribute slots, but just that the parser drops them in that position, so this comment is out of date.

File: lib/Normalize_std_ast.ml [hunk 10]
```diff
@@ -392,20 +408,34 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
     Ast_mapper.default_mapper.typ m typ
   in
   let structure =
+    let structure m str =
+      List.filter str ~f:(fun stri ->
+          match Jane_syntax.Structure_item.of_ast stri with
+          | Some (Jstr_layout (Lstr_kind_abbrev _)) -> false
+          | _ -> true )
+      |> Ast_mapper.default_mapper.structure m
+    in
```
Comment (ccasin) [inline lib/Normalize_std_ast.ml:418]: This, and the similar thing just below, seem surprising to me.  Do I understand correctly we're just filtering out kind abbrevs entirely before doing the round-trip check?

**Review by ccasin — COMMENTED** [review]
Looks mostly good, nice tests, some questions and suggestions below.

File: lib/Ast.ml (hunk 6 above)
Comment (ccasin) [reply lib/Ast.ml:1848]: (Having read the rest of the code, I think the answer is their parenthesization needs are explicitly spelled out in `Fmt_ast.fmt_jkind`)

File: lib/Normalize_std_ast.ml (hunk 10 above)
Comment (dvulakh) [reply lib/Normalize_std_ast.ml:418]: You're right, we should only be filtering them when we're erasing jane syntax (fixed)

File: vendor/parser-jane/repatch.sh (hunk 3 above)
Comment (dvulakh) [reply vendor/parser-jane/repatch.sh]: I commited this change by mistake. I'm reverting `repatch.sh` and will discuss making this change properly with Thomas.

File: test/passing/tests/layout_annotation_mod.ml.js-ref (hunk 1 above)
Comment (dvulakh) [reply test/passing/tests/layout_annotation_mod.ml.js-ref]: FWIW, this looks like an [existing problem](https://github.com/janestreet/ocamlformat/blob/363a1a0f63143bc68066ed8dea20393f5ff76db5/test/passing/tests/layout_annotation.ml.js-ref#L288-L290), but I'll try to see whether it's easy to fix.

Comment (dvulakh) [reply test/passing/tests/layout_annotation_mod.ml.js-ref]: Fixed (and updated internal tests)

File: lib/Ast.ml (hunk 6 above)
Comment (dvulakh) [reply lib/Ast.ml:1848]: Yes, it seems this function is for helping deal with the complicated precedence hierarchy in the expression language. I don't think we need to sneak information about the kind language into `Prec.t`. And yes, we do handle the parenthesization within the formatting logic. I think this is fine for now, because the parenthesization rules are pretty simple. If the kind language grows substantially, we could think about taking a different approach.

File: lib/Exposed.ml (hunk 7 above)
Comment (dvulakh) [reply lib/Exposed.ml]: I added the corresponding logic (but haven't tried writing a file where `ocamlformat` would have produced verboten sequences before the fix).

Comment (dvulakh) [reply lib/Exposed.ml]: It occurred to me also that we should check for e.g.
```ocaml
type t : kind_of < how do I write a class again? >
```
(and have done so)

File: test/passing/tests/layout_annotation_mod.ml.js-ref (hunk 1 above)
Comment (ccasin) [reply test/passing/tests/layout_annotation_mod.ml.js-ref]: Thanks!  (For future reference, "this is an existing problem" is an acceptable complete answer)

**Review by ccasin — APPROVED** [review]
LGTM - merge when you're ready
